---
title: "Pluripotent Proteome"
output:
  html_document:
    toc: true
    toc_depth: 4
    toc_float: 
      collapsed: false
      smooth_scroll: false
    df_print: paged
    code_folding: hide
---

```{r load_packages, warning=FALSE, message=FALSE}
# options
options(stringsAsFactors = F)
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
knitr::opts_knit$set(progress = FALSE)
# packages
# qtl mapping + mediation
library(intermediate) # "simecek/intermediate"
# library(intermediate2) # https://github.com/duytpm16/intermediate2
library(qtl2) 
# # plotting
library(plotly)
library(ggpubr)
library(ggraph)
library(pheatmap)
library(cowplot)
library(ggbeeswarm)
library(GGally)
library(corrplot)
# annotations + general genomic things
#library(biomaRt)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(org.Mm.eg.db)
library(ChIPseeker)
library(GenomicRanges)
#library(ensimplR) # https://github.com/churchill-lab/ensimplR
library(qvalue)
library(LOLA)  

# data processing
library(pcaMethods) # pca
library(Hmisc) # rcorr
# library(WGCNA)
library(gprofiler2)
# library(sva)
library(WebGestaltR)
library(readxl)
library(tidyverse)
select <- dplyr::select # I am adding this explicitly
rename <- dplyr::rename # I am adding this explicitly
# setting path
library(here)
# get functions
source(here("_src/functions.R")) # source all the common functions
```



```{r load_annotate_omics_data, warning=FALSE, message=FALSE, results="hide" }
##### Protein data
# Note the ensembl protein id's are from v98
load(here("../pQTL_website/_data/DO_mESC_pQTL_forMapping_noPoly_v3.RData"))
exprZ.esc_prot <- exprZ
probs.esc_prot <- genoprobs
covar.esc_prot <- covar
covarTidy.esc_prot <- covarTidy
kinship_loco.esc_prot <- kinship_loco
# I don't want to use the rankZ normalized version. let's clean up the raw data for use.
raw.expr.esc_prot <- raw.expr
expr.esc_prot <- raw.expr[rownames(raw.expr) %in% rownames(exprZ), colnames(raw.expr) %in% colnames(exprZ)]
expr.esc_prot <- as.matrix(expr.esc_prot)
rm(exprZ, covar, covarTidy, kinship_loco, genoprobs, raw.expr)
##### RNA data
# Note the ensembl gene id's are from v82
# Note that there are 20 ensembl gene id's that don't match anything even in v82.
load(here("../pQTL_website/_data/DO_mESC_paired_eQTL_forMapping.RData"))
raw.expr.esc_rna <- esc.raw.expr
exprZ.esc_rna <- esc.exprZ
kinship_loco.esc_rna <- esc.kinship_loco
probs.esc_rna <- esc.probs
covar.esc_rna <- esc.covar
covarTidy.esc_rna <- covarTidy
exprComBat.esc_rna <- esc.expr.ComBat
expr.esc_rna <- expm1(exprComBat.esc_rna) # re-transforming since the data was log(x+1) before combat
expr.esc_rna[expr.esc_rna < 0] <- 0
expr.esc_rna <- t(expr.esc_rna)
rm(esc.expr, esc.exprZ, esc.kinship_loco, esc.probs, esc.expr.ComBat, esc.raw.expr, covarTidy, exprComBat.esc_rna, esc.covar, esc.covarTidy)
##### ATAC data
# load the atac peaks
load(here("../pQTL_website/_data/DO_atacseq_prelim_DS0420.RData")) # counts.norm
# remove bad samples
counts.norm <- counts.norm[, !colnames(counts.norm) %in% c("PB357.01", "PB357.17")]
covar.esc_atac <- covar.mat[!rownames(covar.mat) %in% c("PB357.01", "PB357.17"), ]
covarTidy.esc_atac <- covar_full[!covar_full$PB_ID %in% c("PB357.01", "PB357.17"), ]
probs.esc_atac <- probs
# note that there are 2 plates of ATAC samples that have been processed. I will use comBat to remove this effect.
dat <- log1p((counts.norm))
mod <- model.matrix(~sex, data = covarTidy.esc_atac)
exprComBat <- sva::ComBat(
  dat = dat, batch = covarTidy.esc_atac$plate, mod = mod,
  par.prior = TRUE, prior.plots = FALSE
)
counts.norm2 <- expm1(exprComBat)
counts.norm2[counts.norm2 < 0] <- 0
covar.esc_atac <- covar.esc_atac[, "sex", drop = FALSE]
counts.normZ <- apply(counts.norm2, 1, rankZ)
kinship_loco.esc_atac <- calc_kinship(probs.esc_atac, type = "loco")
rm(probs, covar_full, covar.mat, map_dat, dat, mod, exprComBat, counts.norm)
### Annotating using ChipSeeker + TxDb.Mmusculus.UCSC.mm10.knownGene (this uses UCSC data from Oct, 2019 which I assume matches to ensembl v98 (Sep19))
# annotate Atac data
# prep & annotate data
atac.counts <- counts.norm2 %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  separate(rowname, into = c("Chr", "Start", "End"), sep = "_", remove = FALSE) %>%
  mutate(Chr = gsub("peak", "", Chr)) %>%
  column_to_rownames()
gr <- makeGRangesFromDataFrame(atac.counts, 
                               keep.extra.columns = T, 
                               seqnames.field = c("Chr"), 
                               start.field = "Start", 
                               end.field = "End")
gr <- diffloop::addchr(gr)
peakAnno <- annotatePeak(gr,
  TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene, annoDb = "org.Mm.eg.db",
  tssRegion = c(-5000, 5000),
  level = "transcript", assignGenomicAnnotation = TRUE,
  genomicAnnotationPriority = c("Promoter", "5UTR", "3UTR", "Exon", "Intron", "Downstream", "Intergenic"),
  addFlankGeneInfo = FALSE,
  overlap = "TSS", verbose = FALSE
)
atac.peak.annots_full <- as_tibble(peakAnno) %>%
  select(!starts_with("PB")) %>%
  mutate(
    seqnames = gsub("chr", "peak", seqnames),
    chr = gsub("peak", "", seqnames)
  ) %>%
  unite("peak_id", "seqnames", "start", "end", remove = FALSE) %>%
  select(-seqnames)
```

```{r id_matching, message=FALSE, warning=FALSE}
# I am using v98 annotations since protein id's are v98
# Saved all gene annotations from biomart using Sept2019 archive http://sep2019.archive.ensembl.org/biomart
all_annot_v98 <- read_tsv( here("../pQTL_website/_data","ensembl_gene_annotations_v98.txt")) %>% 
rename( "ensembl_gene_id" = "Gene stable ID",
          "protein_id" = "Protein stable ID",
          "gene_start" = "Gene start (bp)",
          "gene_end" = "Gene end (bp)",
          "gene_chr" = "Chromosome/scaffold name",
          "mgi_symbol" = "MGI symbol",
          "gene_biotype" = "Gene type") %>% 
  group_by(ensembl_gene_id) %>% 
  mutate( n_gene_id = n()) %>% # there are cases where gene id is repeated with a valid protein id + NA
  ungroup() %>% 
  filter( !(n_gene_id >1 & is.na(protein_id)) ) %>% 
  select(-n_gene_id)
# make all.genes, all.prots, atac.peak.annots
all.genes <- all_annot_v98 %>% 
  filter( ensembl_gene_id %in% colnames(expr.esc_rna)) %>% 
  select( -protein_id) %>% 
  distinct() 
# all genes:
# 14405/14547 has annotations in v98, 141 missing annotations
# closer look at some show that the ensembl gene ids have been changed. Not sure what the best way in updating those are. 
# Ex: ENSMUSG00000029333 matches Rasgef1b in MGI which has ENSMUSG00000089809 as the new ensembl gene id in v98. I can't find the connection other than manually searching for it! 
  
  
all.prots <- all_annot_v98 %>%
  filter( protein_id %in% colnames(expr.esc_prot)) %>% 
  filter( !is.na(mgi_symbol) ) # I am removing the ones without an annotation 
# 7432/7432 has annotations!
all_atac_peak_gene_annots <- all_annot_v98 %>%  
  filter( ensembl_gene_id %in% atac.peak.annots_full$ENSEMBL) %>% 
  select( -protein_id) %>% 
  distinct() %>% 
  filter( !is.na(mgi_symbol) | !is.na(ensembl_gene_id)) # I am removing the ones without an annotation (n = 3)
# re-writing the annotations. 
atac.peak.annots_full <- atac.peak.annots_full %>% 
  select(peak_id, 
         peak_start = start, 
         peak_end = end, 
         peak_width = width, 
         peak_strand = strand, 
         annotation, 
         distanceToTSS, 
         "ensembl_gene_id"="ENSEMBL") %>% 
  left_join( all_atac_peak_gene_annots)
atac.peak.annots <- atac.peak.annots_full %>% 
  filter( !is.na(mgi_symbol))
# 100430/102173 has annotations
# df with all ensembl gene ids /protein ids found in the three data sets. 
all_omics_ids <- full_join( all_atac_peak_gene_annots, all.genes) %>% 
  full_join( all.prots) #%>% 
  #left_join( select(atac.peak.annots, ensembl_gene_id, peak_id)) # I can add peak_ids if I want to! 
# saving for later use
# save( all.genes, all.prots, atac.peak.annots, file = here("_data","updated_annotations_04042022.RData"))
```


```{r load_qtl_data, warning=FALSE, message=FALSE, results="hide" }

######### merged caQTL, eQTL, pQTL
# see "_src/" for details on how they are merged
load(here("../pQTL_website/_data/peaks_comparison_10Mb_noPoly_v4_w_atac.RData")) # peaks.esc.overlap2, peaks.esc.prot.rna, peaks.esc.rna.atac

######### pQTL data with effects
load(here("../pQTL_website/_data/DO_mESC_pQTL_effects_noPoly_v2.RData"))
peaks.esc_prot <- peaks # all pQTL peaks
effects_blup.esc_prot <- effects_blup # all pQTL effects using blup
rm(peaks, effects_blup, effects_std)
# adding effects to pQTL peaks
peaks.esc_prot.blup <- cbind(peaks.esc_prot, effects_blup.esc_prot) %>% 
  dplyr::rename("protein_id" = "phenotype")
colnames(peaks.esc_prot.blup) <- c(
  colnames(peaks.esc_prot.blup)[1:2],
  paste0(colnames(peaks.esc_prot.blup)[3:dim(peaks.esc_prot.blup)[2]], ".esc_prot")
)
# adding annotations to pQTL with effects
peaks.esc_prot.wEffs <- peaks.esc_prot.blup %>% 
  left_join( all.prots) %>% 
  mutate(midpoint = (gene_start + gene_end) / 2) %>% 
  mutate( same_chrom =  (peak_chr == gene_chr),
          diff = abs(midpoint - interp_bp_peak.esc_prot)) %>% 
  mutate( local.esc_prot = ifelse( same_chrom & 
                            diff < 10e06, TRUE, FALSE
    )) 
# adding RNA + protein details to the effects data frame for all eQTL & pQTL peaks.
peaks.esc_prot.blup <- peaks.esc_prot.blup %>% 
  left_join(., mutate(peaks.esc.prot.rna,
                      peak_cM.esc_prot = as.numeric(peak_cM.esc_prot),
                      peak_cM.esc_rna = as.numeric(peak_cM.esc_rna)
))
######### eQTL data with effects
load(here("../pQTL_website/_data/ESC_eQTL_effects.RData"))
peaks.esc_rna <- peaks
effects_blup.esc_rna <- effects_blup
effects_std.esc_rna <- effects_std
rm(peaks, effects_blup, effects_std)
# Adding effects to eQTL peaks
peaks.esc_rna.blup <- cbind(peaks.esc_rna, effects_blup.esc_rna) %>%
  dplyr::rename("ensembl_gene_id" = "phenotype")
colnames(peaks.esc_rna.blup) <- c(
  colnames(peaks.esc_rna.blup)[1:2],
  paste0(colnames(peaks.esc_rna.blup)[3:dim(peaks.esc_rna.blup)[2]], ".esc_rna")
)
# adding annotations to eQTL peaks with effects
peaks.esc_rna.wEffs <- peaks.esc_rna.blup %>% 
  left_join( all.genes) %>% 
  mutate(midpoint = (gene_start + gene_end) / 2) %>% 
  mutate( same_chrom =  (peak_chr == gene_chr),
          diff = abs(midpoint - interp_bp_peak.esc_rna)) %>% 
  mutate( local.esc_rna = ifelse( same_chrom & 
                            diff < 10e06, TRUE, FALSE
    ))
# adding RNA + protein details to the effects data frame for all eQTL & pQTL peaks.
peaks.esc_rna.blup <- peaks.esc_rna.blup %>% # 
  left_join(., mutate(peaks.esc.prot.rna,
  peak_cM.esc_rna = as.numeric(peak_cM.esc_rna),
  peak_cM.esc_prot = as.numeric(peak_cM.esc_prot)
))
######### caQTL data
load(here("../pQTL_website/_data/atac_effects_at_qtl_peaks.RData")) # peaks
atac.eff_blup <- cbind(peaks, effects_blup)
peaks.esc_atac <- peaks
rm(peaks, effects_blup, effects_std)
# renaming effects
atac.eff_blup %>% 
  select( peak_id = phenotype, peak_chr, lod.esc_atac = lod, interp_bp_peak.esc_atac = interp_bp_peak,
          A.esc_atac = A, 
          B.esc_atac = B,
          C.esc_atac = C,
          D.esc_atac = D, 
          E.esc_atac = E, 
          F.esc_atac = `F`, 
          G.esc_atac = G,
          H.esc_atac = H
          ) -> peaks.esc_atac.blup
# add local distant to peaks.esc_atac.blup
peaks.esc_atac.blup %>% 
  left_join( ., atac.peak.annots_full %>% 
                    mutate(midpoint = (peak_start+peak_end)/2,
                           atac_chr = gsub("_","",substr(peak_id,5,6))) %>% 
                    select( peak_id, midpoint, atac_chr)
             ) %>%  # adding location information for each peak
  mutate( 
    local.esc_atac = (peak_chr == atac_chr & abs( midpoint - as.numeric(interp_bp_peak.esc_atac)) < 10e06 ),
    ) -> peaks.esc_atac.blup
# fixing some probs -- for association mapping 
attributes(probs.esc_rna)$is_x_chr <-  attributes(probs.esc_atac)$is_x_chr
attributes(probs.esc_prot)$is_x_chr <-  attributes(probs.esc_atac)$is_x_chr
#### merge effects
peaks.merged.blup <- full_join( (peaks.esc_rna.blup %>%  select(-bp_diff, -same_chrom) ),
                                 (peaks.esc_prot.blup %>% select(-bp_diff, -same_chrom)) 
                                )
peaks.esc.overlap.wEffs <- peaks.esc.overlap2 %>% 
  mutate( 
    peak_cM.esc_rna = as.numeric(peak_cM.esc_rna),
    peak_cM.esc_prot = as.numeric(peak_cM.esc_prot),
    before.esc_atac = before, 
    after.esc_atac = after
    ) %>% 
  left_join( peaks.esc_atac.blup %>%  select(-midpoint)) %>% 
  left_join( (peaks.esc_rna.wEffs %>%  select( ensembl_gene_id, peak_chr, lod.esc_rna , before.esc_rna, after.esc_rna, paste0(LETTERS[1:8],".esc_rna"))) ) %>% 
  left_join( (peaks.esc_prot.wEffs %>% select( protein_id, peak_chr, lod.esc_prot , before.esc_prot, after.esc_prot, paste0(LETTERS[1:8],".esc_prot"))) )
```


```{r shared_data, warning=FALSE, message=FALSE}
#### Shared data
# get the set of shared genes btw protein + rna
shared.genes <- inner_join( all.genes, all.prots)
# shared in all three data sets
threeway.shared.genes <- atac.peak.annots %>% 
  filter( ensembl_gene_id %in% shared.genes$ensembl_gene_id) %>%  # get genes found in all three omics
  left_join( all.prots) %>% 
  group_by(ensembl_gene_id) %>%
  mutate(new_symbol=paste0(mgi_symbol,"_",1:n()), new_gene_id=paste0(ensembl_gene_id,"_",1:n())) %>% 
  ungroup()
# get the set of shared samples
shared.samples <- intersect(
  rownames(expr.esc_rna)[!grepl("repB", rownames(expr.esc_rna))],
  rownames(expr.esc_prot)[!grepl("repB", rownames(expr.esc_prot))]
)
shared.probs.esc_prot <- probs.esc_prot[ind = shared.samples]
shared.probs.esc_rna <- probs.esc_rna[ind = shared.samples]
# load sample matching table for atac
# load matching sample names for atac-seq data
all.ids <- read_tsv(here("../pQTL_website/_data/Corrected_PBID_SampleMatch_Key_v4.tsv")) %>%
  mutate(sampleid.esc_prot = gsub("rep1", "repA", sampleid.esc_prot))
correct.atac.ids <- read_tsv(here("../pQTL_website/_data/ATAC_sampleids_corrected.tsv"))
sample.matches <- correct.atac.ids %>%
  mutate(sampleid = ifelse(endsWith(correct_id, "B"), paste0(correct_id, "_repB"), paste0(correct_id, "_repA"))) %>%
  mutate(sampleid = ifelse(endsWith(correct_id, "C"), paste0(correct_id, "_repC"), sampleid)) %>%
  full_join(., tibble(rna_id = rownames(expr.esc_rna)), by = c("sampleid" = "rna_id")) %>%
  full_join(., tibble(prot_id = rownames(expr.esc_prot)), by = c("sampleid" = "prot_id"))
sample.matches.nodups <- filter(sample.matches, !endsWith(sampleid, "B"), !endsWith(sampleid, "C"))
# there are some rna/protein samples that are not found in the atac data
# I am adding the top_muga ids for those
missing.samples <- filter(
  all.ids,
  (sampleid.esc_rna %in% filter(sample.matches.nodups, is.na(top_muga))$sampleid & mixup.esc_rna == FALSE) |
    (sampleid.esc_prot %in% filter(sample.matches.nodups, is.na(top_muga))$sampleid & mixup.esc_prot == FALSE)
) %>%
  select(top_muga, sampleid.esc_rna, sampleid.esc_prot, PBID) %>%
  mutate(sampleid = ifelse(is.na(sampleid.esc_rna), sampleid.esc_prot, sampleid.esc_rna)) %>%
  select(-sampleid.esc_rna, -sampleid.esc_prot) %>%
  distinct()
sample.matches.all <- sample.matches.nodups %>%
  full_join(., missing.samples, by = c("sampleid")) %>%
  mutate(
    top_muga = ifelse(!is.na(top_muga.x), top_muga.x, top_muga.y),
    PBID = ifelse(!is.na(PBID.x), PBID.x, PBID.y)
  ) %>%
  select(sampleid, correct_id, ATAC, top_muga, PBID) %>%
  filter(!is.na(top_muga)) # remove the ones that are still NA +  add them back below
# there are still 2 samples that are mismatches with missing top_muga ids
# PB359.33_repA and PB361.72_repA, I am changing those manually using all.ids
mixup.samples <- tibble(
  sampleid = c("PB359.33_repA", "PB361.72_repA"),
  top_muga = c("A903DI1B.C03", "A903DI12.E09"),
  PBID = c("PB359.33", "PB361.72"),
  ATAC = c(NA, NA), correct_id = c(NA, NA)
)
sample.matches.all <- sample.matches.all %>%
  rbind(., mixup.samples)
# For mapping later
# I have factors for all 200 lines + can map with all
# I need to get the genoprobs for all!
# load all 3 and subset + merge
atac.samples <- sample.matches.all %>% 
  filter(ATAC %in% colnames(counts.norm2))
other.samples <- sample.matches.all %>% filter(
  sampleid %in% rownames(expr.esc_rna) |
    sampleid %in% rownames(expr.esc_prot),
  !ATAC %in% colnames(counts.norm2)
)
sample.matches.rna <- sample.matches.all %>%
  filter(sampleid %in% rownames(expr.esc_rna), !is.na(ATAC))
sample.matches.prot <- sample.matches.all %>%
  filter(sampleid %in% rownames(expr.esc_prot), !is.na(ATAC))
# Sharing and setting names
sub.probs.esc_prot <- probs.esc_prot[ind = other.samples$sampleid]
new.ids <- other.samples$top_muga
names(new.ids) <- other.samples$sampleid
sub.probs.esc_prot <- replace_ids(sub.probs.esc_prot, ids = new.ids)
sub.probs.esc_atac <- probs.esc_atac[ind = atac.samples$ATAC]
new.ids2 <- atac.samples$top_muga
names(new.ids2) <- atac.samples$ATAC
sub.probs.esc_atac <- replace_ids(sub.probs.esc_atac, ids = new.ids2)
attributes(sub.probs.esc_atac)$crosstype <- "DO"
attributes(sub.probs.esc_atac)$is_x_chr <- attributes(sub.probs.esc_prot)$is_x_chr
merged.probs <- rbind(sub.probs.esc_atac, sub.probs.esc_prot)
kinship <- calc_kinship(merged.probs, type = "loco")
other.covar <- covar.esc_prot %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  filter(rowname %in% other.samples$sampleid) %>%
  left_join(other.samples, by = c("rowname" = "sampleid")) %>%
  column_to_rownames("top_muga") %>%
  select(sex) %>%
  as.matrix()
atac.covar <- covarTidy.esc_atac[covarTidy.esc_atac$PB_ID %in% atac.samples$ATAC, c("PB_ID", "sex")] %>%
  left_join(atac.samples, by = c("PB_ID" = "ATAC")) %>%
  mutate(sex = ifelse(sex == "F", 0, 1)) %>%
  column_to_rownames("top_muga") %>%
  select(sex) %>%
  as.matrix()
merged.covar <- rbind(atac.covar, other.covar)
threeway.shared.samples <- filter(sample.matches, 
                                  sampleid %in% shared.samples, 
                                  ATAC %in% colnames(atac.counts)) %>%
  left_join(., sample.matches.all)
threeway.shared.probs <- merged.probs[ind = threeway.shared.samples$top_muga]
shared.atac.data <- t(counts.norm2[threeway.shared.genes$peak_id, threeway.shared.samples$ATAC])
colnames(shared.atac.data) <- threeway.shared.genes$new_symbol
rownames(shared.atac.data) <- threeway.shared.samples$sampleid
shared.rna.data <- (expr.esc_rna[threeway.shared.samples$sampleid, threeway.shared.genes$ensembl_gene_id])
colnames(shared.rna.data) <- threeway.shared.genes$new_symbol
shared.prot.data <- (expr.esc_prot[threeway.shared.samples$sampleid, threeway.shared.genes$protein_id])
colnames(shared.prot.data) <- threeway.shared.genes$new_symbol
# get shared qtl peaks
peaks.shared.genes <- peaks.esc.overlap2 %>%
  filter(ensembl_gene_id %in% shared.genes$ensembl_gene_id |
   protein_id %in% shared.genes$protein_id) %>%
  mutate(
    lod.esc_rna = ifelse(is.na(lod.esc_rna), 0, lod.esc_rna),
    lod.esc_prot = ifelse(is.na(lod.esc_prot), 0, lod.esc_prot)
  ) %>%
  mutate(peak_cM.esc_prot = as.numeric(peak_cM.esc_prot), 
         peak_cM.esc_rna = as.numeric(peak_cM.esc_rna))
```

```{r load_gene_sets, warning=FALSE, message=FALSE, results='hide'}
# Complex members retreieved from Ori et al 2016, additional file2:  https://variablecomplexes.embl.de/html-tables/Additional_file_2.html
complexes <- read_xlsx(path = here("../pQTL_website/_data", "protein_complexes.xlsx")) # Assuming they are using 2016 human ensembl id's. but not worrying about it atm.
complex.genes <- (str_split(complexes$`Member identifiers (human Ensembl gene)`, " "))
names(complex.genes) <- complexes$`Complex Name`
# # running the code below and saving to RData file to re-use
# library(biomaRt)
# mart_human <- useMart("ensembl", dataset = "hsapiens_gene_ensembl", host = "http://sep2019.archive.ensembl.org")
# mart_mouse <- useMart("ensembl", dataset = "mmusculus_gene_ensembl", host = "http://sep2019.archive.ensembl.org")
# matched.ids <- getLDS(
#   attributes = c("ensembl_gene_id"),
#   filters = "ensembl_gene_id",
#   values = unique(unlist(complex.genes)),
#   mart = mart_human,
#   attributesL = c("ensembl_gene_id"),
#   martL = mart_mouse
# )
# complex.gene.list <- matched.ids %>%
#   rename( human_ids = `Gene.stable.ID`,
#           ensembl_gene_id = `Gene.stable.ID.1`) %>%
#   left_join(
#     full_join( all.genes, all.prots)
#     ) %>%  # adding all annotations
#   filter( ensembl_gene_id %in% all.genes$ensembl_gene_id |
#             protein_id %in% all.prots$protein_id) # filtering for the genes + proteins in our dataset
# # note that 20 human ids don't have 1-1 mapping, they match 10 genes. Mainly because that complex subunit is not found in mice.
# # complex.gene.list %>%  group_by(ensembl_gene_id, protein_id) %>%  mutate( n =n()) %>% filter( n>1) %>% view()
# 
# save(complex.gene.list, file = here("_data","complex_gene_list.RData"))
load(here("../pQTL_website/_data/complex_gene_list.RData"))
# getting TF list
all.tfs <- readxl::read_xlsx(path = here("../pQTL_website/_data/Mouse_TFs_TcoFDB.xlsx")) %>%
  select(Symbol) %>%
  rename(mgi_symbol = Symbol)
mrna <- expr.esc_rna %>%
  as.data.frame() %>%
  summarize_all(.funs = mean, na.rm = T) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  dplyr::rename(ensembl_gene_id = rowname, avg_rna = V1) %>%
  filter(ensembl_gene_id %in% (filter(all.genes, gene_biotype == "protein_coding"))$ensembl_gene_id) 
  
# esc gene lists
# load lists
# get gene lists for heatmaps
# From Xu et al 2014 paper
# I fed the list to MGI and only Tcf3 has an old symbol Tcf7l1
xu.genes <- read_xlsx(here("../pQTL_website/_data/GeneList-EScell-PluMaint.xlsx"), sheet = 1) %>%
   mutate( Gene = ifelse( Gene =="Tcf3", "Tcf7l1",Gene)) %>% # updating some names to match our mgi_symbols
  left_join(., all_omics_ids, by = c("Gene" = "mgi_symbol")) # Label = "pluripotency", "mesoderm" ...etc.
# From Kalkan et al 2017 paper
# I fed the list to MGI and only Tcf3 has an old symbol Tcf7l1
kalk.genes <- read_xlsx(here("../pQTL_website/_data/GeneList-EScell-Kalkan2017.xlsx"), sheet = 1) %>%
  mutate( Gene = ifelse( Gene =="Tcf3", "Tcf7l1",Gene)) %>% # updating some names to match our mgi_symbols
  left_join(., all_omics_ids, by = c("Gene" = "mgi_symbol")) # Label = " naive_pluripotency" ... etc.
# These are :Potential pluripotency associated genes from multiple genome-wide RNAi screens studies.
# downloaded from: http://www.maayanlab.net/ESCAPE/download/RNAihits.txt.zip
# I fed the list to MGI and many genes have an old symbol, I added those to the file as well - RNAhits_updated.xlsx
# Here merging that to the old list to include both old + new symbols in case we have the older ones in our annotations
escape.genes <- read_tsv(here("../pQTL_website/_data/RNAihits.txt")) %>% # these have capital letter gene names
  select(geneName) %>%
  rbind( read_xlsx(here("../pQTL_website/_data/RNAihits_updated.xlsx"), sheet = 1) %>% 
           select(geneName = Symbol) %>% 
           filter(!is.na(geneName))
  ) %>% 
  distinct() %>% 
  mutate(mgi_symbol = paste0((substring(geneName, 1, 1)), tolower(substring(geneName, 2)))) %>%
  select(-geneName) %>%
  left_join( all_omics_ids) 
# These are: Lists of ESC and diffferentiating ESC-specific proteins. We collected 19,801 ESC and diffferentiating ESC-specific proteins from proteomics. 
# downloaded from: http://www.maayanlab.net/ESCAPE/download/proteomicsESC.txt.zip
# url <- "http://www.maayanlab.net/ESCAPE/download/proteomicsESC.txt.zip"
# download.file(url, here("_data/proteomicsESC.txt.zip"))
# zip::unzip(here("_data/proteomicsESC.txt.zip"), overwrite = F, exdir = "_data/")
# I uploaded the mouse subsets to bimart + converted them manually
escape_diff_genes <- read_tsv( "../pQTL_website/_data/escape_diff_genes.tsv") %>% 
  rename( ensembl_gene_id = `Gene stable ID`, mgi_symbol = `MGI symbol`,geneID = `NCBI gene (formerly Entrezgene) ID`) %>% 
  left_join( ., read_tsv( here("../pQTL_website/_data/proteomicsESC.txt") ) , by = c("geneID" ) ) %>% 
  left_join( all_omics_ids) 
  
dan.genes <- data.frame(
  mgi_symbol = c(
    "Esrrb", "Klf4", "Nanog", "Pou5f1", "Sox2",
    "Chrd", "Fgf8", "Otx2", "Pax6", "Sox1",
    "Gata4","Gata6", "Pdgfra", "Sox17", "Sox7", 
    "Bmp2","Kdr", "Mixl1", "T", "Wnt3a"
  ),
  type = c(rep("pluripotency", 5), rep("ectoderm", 5), rep("endoderm", 5), rep("mesoderm", 5))
) %>%
  left_join(., all_omics_ids)
ortmann_genes <- tibble(
  type = c( rep("primitive endoderm",4),
                  "neurectoderm", "definitive endoderm", "ectoderm","epiblast",
                  rep("Wnt-related genes",8),
            rep("pluripotency",8)),
  mgi_symbol = c( "Gata6","Pdgfra","Sox7","Sparc1",
                  "Sox1","Sox17","Otx2","Fgf5", 
            "Axin2","Lef1","Csnk1e", "Gnai1", "Nkd1", "Tcf7", "Wwnt5b", "Fgfrl1",
            "Pou5f1","Nanog","Klf2","Klf4","Dppa5","Esrrb","Tfcp2l1","Sox2")
)
all.esc.genes <- tibble(mgi_symbol = 
                          unique(c(escape.genes$mgi_symbol, 
                                   xu.genes$Gene,
                                   kalk.genes$Gene,
                                   "Etv5","Rbpj","Fgf15","Dusp9","Id1","Id3","Id4")) # adding some from other papers
                        ) %>%
  left_join(all_omics_ids) %>%
  mutate(s_Mbp = (gene_start / 1e06) - 0.002, e_Mbp = (gene_end / 1e06) + 0.002)
# downloading the ccvariants
# download.file("https://ndownloader.figshare.com/files/9746485", "_data/cc_variants.sqlite")
# read in variant definitions
var.types <- read_xlsx(path = here("../pQTL_website/_data/Variant_Definitions.xlsx")) %>%
  rename(type = `SO term`, score = `Functional consequence score`)

bulut.genes <- readxl::read_xlsx(path = here("../pQTL_website/_data/Bulut_2018_TableS6.xlsx")) %>%
  rename(mgi_symbol = OFFICIAL_GENE_SYMBOL) %>%
  select(-Name, -Species) %>%
  rowwise() %>%
  mutate(mgi_symbol = paste0(substr(mgi_symbol, 1, 1), tolower(substr(mgi_symbol, 2, nchar(mgi_symbol))))) %>%
  rbind(tibble(mgi_symbol = c("Tip60", "Ep400", "Kat5", "Kat8", "Ino80", "Ash2l", "Chd1"))) %>% # these are mentioned by name in the papers
  left_join(all_omics_ids)
# 2c-like cell genes from Hendrickson 2017, table S8
# get the ensembl gene ids, they are from december 2011 acc to the table details
# find most recent versions of the ids using ensembl id converter
# Requested ID	Matched ID(s)
genes_2c <- read_xlsx( path = here("../pQTL_website/_data","Hendrickson_2017_tableS8_2clike_gene_list.xlsx"), sheet = 2) %>% 
  select( ensembl_gene_id_dec2011 = `Requested ID`,	
          ensembl_gene_id = `Matched ID(s)`) %>% 
  left_join( all_annot_v98) %>%
  left_join( all_omics_ids) %>% 
  filter( !is.na(mgi_symbol)) 
```


```{r load_mediations, warning=FALSE, message=FALSE}
# load mediation results
load(here("../pQTL_website/_data/DO_mESC_eQTL_protein_mediation_lod6_v3_merged.RData")) # eqtl_rna_meds
load(here("../pQTL_website/_data/DO_mESC_eQTL_RNA_mediation_lod6_v3_merged.RData")) # eqtl_prot_meds
load(here("../pQTL_website/_data/DO_mESC_pQTL_rna_mediation_lod6_noPoly_v4_merged.RData")) # pqtl_rna_meds
load(here("../pQTL_website/_data/DO_mESC_pQTL_protein_mediation_lod6_noPoly_v4.RData")) # results
pqtl_prot_meds <- results
load(here("../pQTL_website/_data/DO_mESC_caQTL_rna_mediation_merged_v2.RData"))
load(here("../pQTL_website/_data/DO_mESC_caQTL_prot_mediation_merged_v2.RData"))
```

```{r misc, warning=FALSE, message=FALSE, results="hide" }
## lifr genotypes
# get_LIFR_genotypes
# using Dan's code to get LIFR genotypes for the full list of animals
probs <- merged.probs
markers <- tibble(name = dimnames(probs[[15]])[[3]]) %>%
  mutate(name2 = name) %>%
  separate(name2, into = c("chrom", "pos"), sep = "_", convert = TRUE)
# LIFr SNP is chr15:7116944 (rs50454566)
mm <- filter(markers, chrom == "15", pos > 7090000, pos < 7130000) # 3 markers
probs2 <- probs$`15`[, , mm$name]
closest_geno <- function(p, tol = 0.01) {
  if (sum(abs(p - c(1, 0))) < tol) {
    return("A")
  }
  if (sum(abs(p - c(0, 1))) < tol) {
    return("B")
  }
  if (sum(abs(p - c(0.5, 0.5))) < tol) {
    return("H")
  }
  return(NA)
}
call_geno <- function(mat) {
  # mat is nsamp*8 (haps)
  # A = A_J
  # B = B6
  # C = 129
  # D = NOD
  # E = NZO
  # F = CAST
  # G = PWK
  # H = WSB
  # I want to divide NOD + CAST + PWK + WSB
  # vs. the other four
  grp <- c("A", "A", "A", "B", "A", "B", "B", "B")
  collapsed <- apply(mat, 1, function(x) tapply(x, grp, sum))
  apply(collapsed, 2, closest_geno)
}
probs3 <- apply(probs2, 3, call_geno)
# assert_that(noNA(probs3))
one <- probs3[, 1] # marker left of Lifr
two <- probs3[, 2] # closest marker to Lifr
three <- probs3[, 3] # marker right of Lifr
# "PB360.49" has an ancestry switch between markers 1 & 2!
# Get samples in group A (inbred strains) and group B (wild-derived + NOD)
inbred <- rownames(probs[[1]])[one == "A" & two == "A" & three == "A"]
wildder <- rownames(probs[[1]])[one == "B" & two == "B" & three == "B"]
hets <- rownames(probs[[1]])[one == "H" & two == "H" & three == "H"]
# cat(inbred, sep="\n", file="lifr_genotype_inbred.txt")
# cat(wildder, sep="\n", file="lifr_genotype_wildder.txt")
# cat(hets, sep="\n", file="lifr_genotype_het.txt")
#
data_frame(
  lifr_geno = factor(c(rep("Ref", length(inbred)), c(rep("Alt", length(wildder))), c(rep("Het", length(hets))))),
  lifr = factor(c(rep(0, length(inbred)), c(rep(1, length(wildder))), c(rep(2, length(hets))))),
  rowname = c((inbred), (wildder), (hets))
) %>%
  mutate(rowname = ifelse(is.na(rowname), "A903DI1C.C12", rowname)) -> covar.lifr
merged.covar %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  full_join(covar.lifr) %>%
  column_to_rownames("rowname") -> merged.covar2
merged.covar3 <- model.matrix(~ sex + lifr, merged.covar2)[, -1]
shared.probs <- merged.probs[ind = threeway.shared.samples$top_muga]
attributes(shared.probs)$is_x_chr <-  attributes(probs.esc_atac)$is_x_chr
shared.covar <- merged.covar[ threeway.shared.samples$top_muga,,drop=FALSE]
shared.kinship <- calc_kinship(shared.probs, type = "loco")
# prep some stuff for plotting:
uchr <- c(as.character(1:19), "X")
cl <- dplyr::select(map_dat2, chr, pos_bp) %>%
  group_by(chr) %>%
  dplyr::summarize(len = max(pos_bp))
clp <- with(cl, setNames(len, chr))
chrom_lens <- setNames(as.numeric(clp[uchr]), uchr)
chrom_lens_offset <- cumsum(chrom_lens) - chrom_lens
chrom_lens_midpt <- chrom_lens_offset + chrom_lens / 2
all.genes2 <- all.genes %>% mutate(midpoint = (gene_start + gene_end) / 2)
all.genes2$cumsum_bp_gene <- all.genes2$midpoint + chrom_lens_offset[all.genes2$gene_chr]
all.prots2 <- all.prots %>% mutate(midpoint = (gene_start + gene_end) / 2)
all.prots2$cumsum_bp_gene <- all.prots2$midpoint + chrom_lens_offset[all.prots2$gene_chr]
map_dat2 <- rename(map_dat2, pos_cM = pos)
map_dat2 <- mutate(map_dat2, pos_cM = as.numeric(pos_cM))

#  rna / prot / chromatin
qtl.colors <- c( rna = "#228833", 
                 prot = "#4477AA", 
                 atac = "#EE6677",
                 shared = "#AA3377")
#founder_colors <- c("#FFDC00", "#888888", "#F08080", "#0064C9", "#7FDBFF", "#2ECC40", "#FF4136", "#B10DC9")
founder_colors <- c(AJ = "#F0E442", B6 = "#555555", `129` = "#E69F00", NOD = "#0072B2",
   NZO = "#56B4E9", CAST = "#009E73", PWK = "#D55E00", WSB = "#CC79A7")
## adding cc_variants for association mapping
query_variants <- create_variant_query_func(here("../pQTL_website/_data/cc_variants.sqlite"))
query_genes <- create_gene_query_func(here("../pQTL_website/_data/mouse_genes_mgi.sqlite"))
# sex calls from Alex Stanton retrieved via slack on July 28, 2020
sex_calls <- read_csv(here("../pQTL_website/_data/ploidystatus_AS_07282020.csv"))
# for LOLA
regionDB = loadRegionDB(here("_data/LOLA/nm/t1/resources/regions/LOLACore/mm10/"))

```


```{r rna_var}
# rna
var.rna <- expr.esc_rna %>%
  as_tibble(.) %>%
  summarise_all(list(~ var(., na.rm = T))) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  rename(ensembl_gene_id = rowname, var = V1) %>%
  arrange(desc(var))
n.rna <- expr.esc_rna %>%
  as_tibble(.) %>%
  summarise_all(list(~ sum(!is.na(.)))) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  rename(ensembl_gene_id = rowname, n = V1) %>%
  arrange(desc(n))
mean.rna <- expr.esc_rna %>%
  as_tibble(.) %>%
  summarise_all(list(~ mean(., na.rm = T))) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  rename(ensembl_gene_id = rowname, mean = V1) %>%
  arrange(desc(mean))
var.rna <- inner_join(var.rna, mean.rna) %>%
  inner_join(., n.rna) %>%
  left_join(., select(all.genes, ensembl_gene_id, ensembl_gene_id, mgi_symbol, gene_chr)) %>%
  mutate(sd = sqrt(var)) %>%
  mutate(cv.rna = 100 * sd / (mean)) %>%
  rename(mean.rna = mean, sd.rna = sd)
```

```{r prot_var}
# protein
var.prot <- expr.esc_prot %>%
  as_tibble(.) %>%
  summarise_all(list(~ var(., na.rm = T))) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  rename(protein_id = rowname, var = V1) %>%
  arrange(desc(var))
n.prot <- expr.esc_prot %>%
  as_tibble(.) %>%
  summarise_all(list(~ sum(!is.na(.)))) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  rename(protein_id = rowname, n = V1) %>%
  arrange(desc(n))
mean.prot <- expr.esc_prot %>%
  as_tibble(.) %>%
  summarise_all(list(~ mean(., na.rm = T))) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  rename(protein_id = rowname, mean = V1) %>%
  arrange(desc(mean))
var.prot <- inner_join(var.prot, mean.prot) %>%
  inner_join(., n.prot) %>%
  left_join(., select(all.prots, protein_id, protein_id, ensembl_gene_id, mgi_symbol, gene_chr)) %>%
  mutate(sd = sqrt(var)) %>%
  mutate(cv.prot = 100 * sd / (mean)) %>%
  rename(mean.prot = mean, sd.prot = sd)
```

```{r atac_rna_var, cache=TRUE}
# atac - filtering for ATACseq peaks in shared genes
var.atac <- atac.counts[threeway.shared.genes$peak_id, 4:ncol(atac.counts)] %>%
  t() %>%
  as_tibble(.) %>%
  summarise_all(list(~ var(., na.rm = T))) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  rename(peak_id = rowname, var = V1) %>%
  arrange(desc(var))
n.atac <- atac.counts[threeway.shared.genes$peak_id, 4:ncol(atac.counts)] %>%
  t() %>%
  as_tibble(.) %>%
  summarise_all(list(~ sum(!is.na(.)))) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  rename(peak_id = rowname, n = V1) %>%
  arrange(desc(n))
mean.atac <- atac.counts[threeway.shared.genes$peak_id, 4:ncol(atac.counts)] %>%
  t() %>%
  as_tibble(.) %>%
  summarise_all(list(~ mean(., na.rm = T))) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  rename(peak_id = rowname, mean = V1) %>%
  arrange(desc(mean))
var.atac <- inner_join(var.atac, mean.atac) %>%
  inner_join(., n.atac) %>%
  left_join(., select(threeway.shared.genes, peak_id, ensembl_gene_id, protein_id, mgi_symbol, gene_chr)) %>%
  mutate(sd = sqrt(var)) %>%
  mutate(cv.atac = 100 * sd / (mean)) %>%
  rename(mean.atac = mean, sd.atac = sd)
all.var <- left_join(
  select(var.atac, peak_id, ensembl_gene_id, protein_id, cv.atac, mean.atac, sd.atac),
  select(var.prot, protein_id, ensembl_gene_id, cv.prot, mean.prot, sd.prot)
) %>%
  left_join(select(var.rna, ensembl_gene_id, cv.rna, mean.rna, sd.rna)) 
```


```{r heritability, cache= TRUE, warning=FALSE, message=FALSE}
kinship.esc_prot <- calc_kinship( probs.esc_prot)
kinship.esc_rna <- calc_kinship( probs.esc_rna)
kinship.esc_atac <- calc_kinship( probs.esc_atac)
herit.esc_prot <- est_herit(
  pheno = exprZ.esc_prot,
  kinship = kinship.esc_prot, 
  addcovar = covar.esc_prot
)
herit.esc_rna <- est_herit(
  pheno = exprZ.esc_rna,
  kinship = kinship.esc_rna, 
  addcovar = covar.esc_rna
)
herit.esc_atac <- est_herit(
  pheno = counts.normZ,
  kinship = kinship.esc_atac, 
  addcovar = covar.esc_atac
)
```


## Figure 1A: Overview of data sets

```{r fig1a}



```

We have a unique data set where 200 DO embryonic stem cell lines with distinct genotypes have chromatin accessibility, transcriptome and proteome measures with 163 of them containing all three. Below is a brief summary of each data, please see the manuscript for details. The goal of this website is to accompany the Aydin et al. 2022 manuscript, providing the necessary data and code for reproducibility.

 
Below are the filtered, normalized abundances of `r formatC(ncol(expr.esc_prot), big.mark=",")` proteins across `r nrow(expr.esc_prot)` DO mESC lines (Table S1 in Aydin et al., (2022)).  
 
```{r protein_abundance_table}

expr.esc_prot %>% 
  as_tibble( rownames = "sampleid") %>% 
  pivot_longer( cols = 2:7433, names_to = "protein_id", values_to = "protein_abundance") %>%
  left_join( all.prots) %>% 
  select( sampleid, mgi_symbol, protein_id, protein_abundance ) %>% 
  pivot_wider( id_cols = c("protein_id","mgi_symbol"), 
               names_from = "sampleid", 
               values_from = "protein_abundance") %>% 
  # mutate_if( is.numeric, formatC, format = "fg", digits =2 ) %>% 
  create_dt()

```

Below are the sample details including the sex and genotype at the *Lifr* locus for each cell line.

```{r protein_samples_table}

covarTidy.esc_prot %>% 
  left_join(covar.lifr, by = c("top_muga"="rowname")) %>% 
  select(sampleid, sex, lifr_geno) %>% 
  mutate( sex = ifelse( sex == "M", "Male", "Female")) %>% 
  rename( `Genotype at Lifr` = lifr_geno) %>% 
  create_dt()

```

 
## Figure 1B: Detection bias in proteomics

**add here:**
  -   Fig1b generation: probablility of detection given transcript abundance. 
  -   Table of values plotted to accompany the figure for download.
  -   ORA for detected + not detected proteins -- Old table S1 - new table S2
  -   Figures S1A, B, C
  -   

## Figure 1C: Principal component analysis of the pluripotent proteome

**add here:**
  -   Fig1C generation: PCA plot
  -   Table of values plotted to accompany the fig for download.
  -   Table showing top drivers of PC1 & statistical test for enrichment of X chr genes
  -   ORA with PC1 drivers - table
  -   Fig S1D, E
  -   sex effects statistical test code + table with results
  
## Figure 1D: Geneset variation analysis of the pluripotent proteome

**add here:**
  -   Fig1D generation: GSVA results
  -   Table of GSVA results for protein + rna with accompanying code showing how it is generated -- old table s2 new table S3
  -   ORA with most/least variable proteins + table
  -   Fig SF, G
  
  
  
  
